### ISSUE TITLE
 Gradient Descent에 대해서 쉽게 설명한다면?

### ISSUE THREAD

**Issue by kimyuuuuuu** [AInsteins/AI-Tech-Interview#34]

**Comment by jarammm** [comment]
> gd란? 1차 미분계수를 이용해 함수의 최소값을 찾아가는 방법입니다.

<미분을 이용한 함수값의 증가 및 감소>
경사상승법 : 미분값을 더하여 함수의 극대값의 위치를 구함
경사하강법 : 미분값을 빼 함수의 극소값의 위치를 구함

<변수가 벡터일 경우>
- 벡터가 입력인 다변수 함수일 경우 편미분(partial differentiation) 사용
- 각 변수 별로 편미분을 계산한 그레디언트 벡터를 이용하여 경사하강 및 경사상승법에 사용할 수 있음
- 그레디언트 벡터 : 하나의 벡터 안의 원소에 대해 각각 미분한 것
- 경사하강법 : 경사하강법 알고리즘을 그대로 적용하되, 벡터는 절대값 대신 노름(norm)을 계산해서 종료조건을 설정

<gd로 벡터 업데이트하는 방법>
- t+1번째 벡터 = t번째 벡터 - 학습률(lambda) * t번째 벡터에서의 미분 벡터

<gd의 한계>
- 미분 가능 + 볼록(convex)한 함수에 대해서는 수렴이 보장되지만, 그렇지 않은 경우 보장 안됨
- 비선형회귀 문제 : non-convex할 수 있음 → 수렴 항상 보장 x
- 특히 딥러닝을 사용하는 경우 목적식의 대부분은 볼록함수가 아님

**Comment by import-max** [comment]
- 어떤 함수의 global minimum을 찾기 위해 기울기를 따라 이동해가는 방법이다. 보통 딥러닝에서는 loss function을 최소화하기 위해 loss function값을 미분하여 기울기를 구하고 경사가 하강하는 방향으로 파라미터 값을 계속 업데이트 해나간다.
- learning rate(학습률)은 한 번에 이동하는 거리의 크기인데, 적절한 크기를 찾는 것이 중요하다. 너무 크거나 너무 작으면 연산 시간이 오래 걸리거나 최소 값을 구하지 못할 수 있다.
- local minima (지역적 국소부위)에 빠지는 경우에 있는데, 보통 momentum 등의 방법을 사용하여 이를 해결할 수 있다.
    - momentum이란 물리학에서 사용하는 단어인데, 어떤 물체가 특정한 방향으로 계속해서 가는 등의 관성을 모멘텀이라고 한다.
    - momentum을 사용하게 되면 일정한 방향을 유지하면서 움직이고 가속도처럼 같은 방향으로 더 많이 변화시키기 때문에 학습속도를 높여준다.
    - [[reference](https://ynebula.tistory.com/26)](https://ynebula.tistory.com/26)

**Comment by kimyuuuuuu** [comment]
- Gradient Descent(경사 하강) 개념
    - 함수 값이 낮아지는 방향으로 독립 변수 값을 변형시켜가면서 최종적으로는 최소 함수 값을 갖도록 하는 독립 변수 값을 찾는 방법
    - 어떤 함수의 극소점을 찾기 위해 gradient(기울기) 반대 방향으로 이동해 가는 방법
    - 딥러닝에서는 **Loss function을 최소화**시키기 위해 파라미터에 대해 Loss function을 미분하여 그 기울기값(gradient)을 구하고, 경사가 하강하는 방향으로 파라미터 값을 점진적으로 찾기위해 사용된다.

- **gradient descent의 목적과 사용 이유**
    - gradient descent는 **함수의 최소값**을 찾는 문제에서 활용된다.
    - 미분계수가 0인 지점을 찾는 방식이 아닌 gradient descent를 이용해 함수의 최소값을 찾는 주된 이유 = 우리가 주로 실제 분석에서 맞딱드리게 되는 함수들은 닫힌 형태(closed form)가 아니거나 함수의 형태가 복잡해 (가령, 비선형함수) 미분계수와 그 근을 계산하기 어려운 경우가 많고,실제 미분계수를 계산하는 과정을 컴퓨터로 구현하는 것에 비해 gradient descent는 컴퓨터로 비교적 쉽게 구현할 수 있기 때문이다.
    - 데이터 양이 매우 큰 경우 gradient descent와 같은 iterative한 방법을 통해 해를 구하면 계산량 측면에서 더 효율적으로 해를 구할 수 있다.

- Gradient Descent의 문제점
    - 적절한 step size(learning rate)가 필요하다.
        - step size가 큰 경우 한 번 이동하는 거리가 커지므로 빠르게 수렴 할 수 있다는 장점이 있다. 하지만, step size를 너무 크게 설정해버리면 최소값을 계산하도록 수렴하지 못하고 함수 값이 계속 커지는 방향으로 최적화가 진행될 수 있다.
        - 한편 step size가 너무 작은 경우 발산하지는 않을 수 있지만 최적의 x를 구하는데 소요되는 시간이 오래 걸린다는 단점이 있다.
        
    - local minima
        - gradient descent 알고리즘을 시작하는 위치는 매번 랜덤하기 때문에 어떤 경우에는 local minima에 빠져 계속 헤어나오지 못하는 경우도 생긴다.
